src/components/scatterplot/ScatterplotSubscriber.js

```diff
@@ -150,13 +151,26 @@ export default function ScatterplotSubscriber(props) {
   }), [cellColorEncoding, geneSelection, mergedCellSets,
     cellSetSelection, cellSetColor, expressionData, attrs]);
 
-  const cellSetPolygons = useMemo(() => getCellSetPolygons({
-    cells,
-    mapping,
-    cellSets: mergedCellSets,
-    cellSetSelection,
-    cellSetColor,
-  }), [cells, mapping, mergedCellSets, cellSetSelection, cellSetColor]);
+  // cellSetPolygonCache is an array of tuples like [(key0, val0), (key1, val1), ...],
+  // where the keys are cellSetSelection arrays.
+  const [cellSetPolygonCache, setCellSetPolygonCache] = useState([]);
+  const cacheHas = (cache, key) => cache.findIndex(el => isEqual(el[0], key)) !== -1;
+  const cacheGet = (cache, key) => cache.find(el => isEqual(el[0], key))?.[1];
+  const cellSetPolygons = useMemo(() => {
+    if (cellSetPolygonsVisible && !cacheHas(cellSetPolygonCache, cellSetSelection)) {
+      const newCellSetPolygons = getCellSetPolygons({
+        cells,
+        mapping,
+        cellSets: mergedCellSets,
+        cellSetSelection,
+        cellSetColor,
+      });
+      setCellSetPolygonCache(cache => [...cache, [cellSetSelection, newCellSetPolygons]]);
+      return newCellSetPolygons;
+    }
+    return cacheGet(cellSetPolygonCache, cellSetSelection);
```

Comment:
```suggestion
    return cacheGet(cellSetPolygonCache, cellSetSelection) || [];
```

The http://localhost:3000/?dataset=dries-2019 demo crashes when `cellSetPolygons` is undefined